c10/core/TensorImpl.h

```diff
@@ -1380,6 +1415,18 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
 
   PyObject* pyobj_ = nullptr; // weak reference
 
+  // An `opaque_handle_` allows customized data layout other than the default strided layout.
+  // Storage is initialized and it is valid to get data pointer and range via `data()` and `numel()`.
+  // Metadata like device, dtype and dims can be queried like a normal tensor with strided layout
+  // but cannot be changed. Therefore `allow_tensor_metadata_change()` always returns
+  // `false` and `set_allow_tensor_metadata_change()` is no-op. The `strides()` is not supported.
+  // Calling any method that changes the metadata and storage would fail. Calling to `is_contiguous()`
+  // always returns false since "contiguous" is not well-defined for a customized layout.
+  //
+  // Following invariant holds when `opaque_handle_` is valid:
+  //     `!is_contiguous() && !allow_tensor_metadata_change()`
+  c10::intrusive_ptr<c10::intrusive_ptr_target> opaque_handle_;
```

Comment:
wait I thought we don't need opque_handle_ in the top level tensor class once we have OpqueTensorImpl?

Reply:
yes, good catch.  I must have messed up the merge, let me try again.